editor/src/components/shared/scoped-variables.ts

```diff
@@ -0,0 +1,146 @@
+  insertableComponent,
+  insertableComponentGroupProjectComponent,
+} from './project-components'
+import { emptyImports } from '../../core/workers/common/project-file-utils'
+
+export function getVariablesInScope(
+  elementPath: ElementPath | null,
+  projectContents: ProjectContentTreeRoot,
+  variablesInScopeFromEditorState: VariablesInScope,
+): AllVariablesInScope[] {
+  return withUnderlyingTarget(
+    elementPath,
+    projectContents,
+    [],
+    (success, element, underlyingTarget, underlyingFilePath) => {
+      let varsInScope = []
+
+      if (elementPath !== null) {
+        const componentScopedVariables = getVariablesFromComponent(
+          success.topLevelElements,
+          elementPath,
+          variablesInScopeFromEditorState,
+        )
+        varsInScope.push(componentScopedVariables)
+      }
+
+      /** for future reference - adding variables from top level **/
+      // const topLevelVariables = getTopLevelVariables(success.topLevelElements, underlyingFilePath)
+      // varsInScope.push(topLevelVariables)
+
+      return varsInScope
+    },
+  )
+}
+
+function getTopLevelVariables(topLevelElements: TopLevelElement[], underlyingFilePath: string) {
+  const topLevelCode: ArbitraryJSBlock[] = topLevelElements.filter((el) =>
+    isArbitraryJSBlock(el),
+  ) as ArbitraryJSBlock[]
+  const topLevelVariables = topLevelCode?.map((block) => block?.definedWithin ?? []).flat()
+
+  return {
+    filePath: underlyingFilePath,
+    variables: topLevelVariables.map((variable) => ({
+      name: variable,
+      type: 'string' as const,
+    })),
+  }
+}
+
+function getVariablesFromComponent(
+  topLevelElements: TopLevelElement[],
+  elementPath: ElementPath,
+  variablesInScopeFromEditorState: VariablesInScope,
+) {
+  const elementPathString = toComponentId(elementPath)
+  const jsxComponent = topLevelElements.find((el) => isUtopiaJSXComponent(el)) as UtopiaJSXComponent
+  const jsxComponentVariables = variablesInScopeFromEditorState[elementPathString] ?? {}
+  return {
+    filePath: jsxComponent?.name ?? 'Component',
+    variables: Object.entries(jsxComponentVariables).map(([name, value]) => ({
+      name,
+      value,
+      type: typeof value,
+    })),
+  }
+}
+
+export function convertVariablesToElements(
+  variableInScope: AllVariablesInScope[],
+): InsertableComponentGroup[] {
+  return variableInScope.map((variableGroup) => {
+    return {
+      source: insertableComponentGroupProjectComponent(variableGroup.filePath),
+      insertableComponents: variableGroup.variables.map((variable) => {
+        return insertableComponent(
+          emptyImports(),
+          () => getMatchingElementForVariable(variable),
+          variable.name,
+          [],
+          null,
+        )
+      }),
+    }
+  })
+}
+
+function getMatchingElementForVariable(variable: Variable) {
+  switch (variable.type) {
+    case 'string':
+      return jsxElementWithoutUID('span', jsxAttributesFromMap({}), [
```

Comment:
writing here for posterity,but we chatted about this. I think it was better when the code used `jsxTextBlock` here, it achieves the exact same outcome but doesn't involve hijacking unrelated types and touching the parser-printer